File: go/test/dns_result_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/whois-domain-monitoring-sdk/go"
	"github.com/voxgig-sdk/whois-domain-monitoring-sdk/go/core"

	vs "github.com/voxgig-sdk/whois-domain-monitoring-sdk/go/utility/struct"
)

func TestDnsResultEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.DnsResult(nil)
		if ent == nil {
			t.Fatal("expected non-nil DnsResultEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := dns_resultBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "dns_result." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		dnsResultRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath(setup.data, "existing.dns_result")))
		var dnsResultRef01Data map[string]any
		if len(dnsResultRef01DataRaw) > 0 {
			dnsResultRef01Data = core.ToMapAny(dnsResultRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = dnsResultRef01Data

		// LOAD
		dnsResultRef01Ent := client.DnsResult(nil)
		dnsResultRef01MatchDt0 := map[string]any{}
		dnsResultRef01DataDt0Loaded, err := dnsResultRef01Ent.Load(dnsResultRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if dnsResultRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func dns_resultBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "dns_result", "DnsResultTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read dns_result test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse dns_result test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"dns_result01", "dns_result02", "dns_result03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID": idmap,
		"WHOIS_DOMAIN_MONITORING_TEST_LIVE":      "FALSE",
		"WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN":   "FALSE",
		"WHOIS_DOMAIN_MONITORING_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["WHOIS_DOMAIN_MONITORING_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewWhoisDomainMonitoringSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/dns_result_entity_test.lua
-- DnsResult entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("whois-domain-monitoring_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("DnsResultEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:DnsResult(nil)
    assert.is_not_nil(ent)
  end)

  it("should run basic flow", function()
    local setup = dns_result_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"load"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "dns_result." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- Bootstrap entity data from existing test data.
    local dns_result_ref01_data_raw = vs.items(helpers.to_map(
      vs.getpath(setup.data, "existing.dns_result")))
    local dns_result_ref01_data = nil
    if #dns_result_ref01_data_raw > 0 then
      dns_result_ref01_data = helpers.to_map(dns_result_ref01_data_raw[1][2])
    end

    -- LOAD
    local dns_result_ref01_ent = client:DnsResult(nil)
    local dns_result_ref01_match_dt0 = {}
    local dns_result_ref01_data_dt0_loaded, err = dns_result_ref01_ent:load(dns_result_ref01_match_dt0, nil)
    assert.is_nil(err)
    assert.is_not_nil(dns_result_ref01_data_dt0_loaded)

  end)
end)

function dns_result_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/dns_result/DnsResultTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read dns_result test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "dns_result01", "dns_result02", "dns_result03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID"] = idmap,
    ["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] = "FALSE",
    ["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] = "FALSE",
    ["WHOIS_DOMAIN_MONITORING_APIKEY"] = "NONE",
  })

  local idmap_resolved = helpers.to_map(
    env["WHOIS_DOMAIN_MONITORING_TEST_DNS_RESULT_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      {
        apikey = env["WHOIS_DOMAIN_MONITORING_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["WHOIS_DOMAIN_MONITORING_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["WHOIS_DOMAIN_MONITORING_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
